include/phasar/DataFlow/IfdsIde/Solver/IDESolverAPIMixin.h

```diff
@@ -0,0 +1,534 @@
+        return finalize();
+      }
+      return std::nullopt;
+    }();
+  }
+
+  /// Continues running the solver on the configured problem after it got
+  /// interrupted in a previous run. Periodically checks whether
+  /// IsCancelled is true.
+  ///
+  /// \remark Please make sure to *only* call this function on an IDESolver
+  /// where the solving process is interrupted, i.e. one of the interruptable
+  /// solving methods returned std::nullopt. It is *invalid* to call this
+  /// function on an IDESolver that has not yet started solving or has already
+  /// flinalized solving.
+  ///
+  /// \returns An std::optional holding a view into the analysis results or
+  /// std::nullopt if the analysis was cancelled.
+  auto continueWithAsyncCancellation(std::atomic_bool &IsCancelled) && {
+    using RetTy =
+        std::optional<std::decay_t<decltype(std::move(*this).finalize())>>;
+    return [&]() -> RetTy {
+      if (continueWithAsyncCancellationImpl(IsCancelled)) {
+        return std::move(*this).finalize();
+      }
+      return std::nullopt;
+    }();
+  }
+
+private:
+  [[nodiscard]] Derived &self() &noexcept {
+    static_assert(std::is_base_of_v<IDESolverAPIMixin, Derived>,
+                  "Invalid CRTP instantiation");
+    return static_cast<Derived &>(*this);
+  }
+
+  void continueImpl() {
+    while (next()) {
+      // no interrupt in normal solving process
+    }
+  }
+
+  void solveImpl() {
+    if (initialize()) {
+      continueImpl();
+    }
+  }
+
+  template <typename CancellationRequest>
+  [[nodiscard]] bool
+  continueUntilImpl(CancellationRequest CancellationRequested,
+                    std::chrono::milliseconds Interval) {
+    auto IsCancellationRequested =
+        [&CancellationRequested](
+            std::chrono::steady_clock::time_point TimeStamp) {
+          if constexpr (std::is_invocable_r_v<
+                            bool, CancellationRequest,
+                            std::chrono::steady_clock::time_point>) {
+            return std::invoke(CancellationRequested, TimeStamp);
+          } else {
+            return std::invoke(CancellationRequested);
+          }
+        };
+
+    size_t NumIterations = Interval.count() * 500;
+
+    auto Start = std::chrono::steady_clock::now();
+
+    while (nextN(NumIterations)) {
+      auto End = std::chrono::steady_clock::now();
+      using milliseconds_d = std::chrono::duration<double, std::milli>;
+
+      auto DeltaTime = std::chrono::duration_cast<milliseconds_d>(End - Start);
+      Start = End;
+
+      if (IsCancellationRequested(End)) {
+        return false;
+      }
+
+      // Adjust NumIterations
+      auto IterationsPerMilli = double(NumIterations) / DeltaTime.count();
+      auto NewNumIterations =
+          size_t(IterationsPerMilli * double(Interval.count()));
+      NumIterations = (NumIterations + 2 * NewNumIterations) / 3;
+    }
+    auto End = std::chrono::steady_clock::now();
+    return !IsCancellationRequested(End);
+  }
```

Comment:
I wonder how robust this is. propagate() is not constant time, but not that diverse in its runtime. Still the question is what is the granularity by which we want to be able to stop. If this here is good enough, okay. But maybe the time checking could/should be done in a thread?